In [1]:
import numpy as np
from numpy import array
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.decomposition import PCA
from numpy.linalg import eig
from sklearn.preprocessing import StandardScaler
import plotly.plotly as py

In [2]:
feature_name = ['Class','Alcohol',  'Malic acid', 'Ash' , 'Alcalinity of ash', 'Magnesium', 'Total phenols', 'Flavanoids',
'Nonflavanoid phenols', 'Proanthocyanins', 'Color intensity', 'Hue', 'OD280/OD315 of diluted wines', 'Proline'] 

In [3]:
raw_data = pd.read_csv('wine.data', names = feature_name)

In [4]:
features = raw_data.iloc[:,1:]

In [5]:
std_scale = StandardScaler()
standard_scaled_data = std_scale.fit_transform(features)
standard_scaled_data[0:5,:]

/Users/aakashvarshney/.local/lib/python3.7/site-packages/sklearn/preprocessing/data.py:625: DataConversionWarning:

Data with input dtype int64, float64 were all converted to float64 by StandardScaler.

/Users/aakashvarshney/.local/lib/python3.7/site-packages/sklearn/base.py:462: DataConversionWarning:

Data with input dtype int64, float64 were all converted to float64 by StandardScaler.



array([[ 1.51861254, -0.5622498 ,  0.23205254, -1.16959318,  1.91390522,
         0.80899739,  1.03481896, -0.65956311,  1.22488398,  0.25171685,
         0.36217728,  1.84791957,  1.01300893],
       [ 0.24628963, -0.49941338, -0.82799632, -2.49084714,  0.01814502,
         0.56864766,  0.73362894, -0.82071924, -0.54472099, -0.29332133,
         0.40605066,  1.1134493 ,  0.96524152],
       [ 0.19687903,  0.02123125,  1.10933436, -0.2687382 ,  0.08835836,
         0.80899739,  1.21553297, -0.49840699,  2.13596773,  0.26901965,
         0.31830389,  0.78858745,  1.39514818],
       [ 1.69154964, -0.34681064,  0.4879264 , -0.80925118,  0.93091845,
         2.49144552,  1.46652465, -0.98187536,  1.03215473,  1.18606801,
        -0.42754369,  1.18407144,  2.33457383],
       [ 0.29570023,  0.22769377,  1.84040254,  0.45194578,  1.28198515,
         0.80899739,  0.66335127,  0.22679555,  0.40140444, -0.31927553,
         0.36217728,  0.44960118, -0.03787401]])

In [6]:
transpose_matrix = standard_scaled_data.T

In [7]:
feature_cov_matrix = np.cov(transpose_matrix)

In [8]:
features_corr_matrix = np.corrcoef(transpose_matrix)

In [9]:
eigenvalues, eigenvectors = eig(feature_cov_matrix)

Singular Vector Decomposition

In [10]:
print ("Eigenvectors", eigenvectors)

Eigenvectors [[-0.1443294   0.48365155 -0.20738262  0.0178563  -0.26566365  0.21353865
   0.05639636 -0.01496997  0.39613926 -0.26628645 -0.50861912 -0.22591696
   0.21160473]
 [ 0.24518758  0.22493093  0.08901289 -0.53689028  0.03521363  0.53681385
  -0.42052391 -0.02596375  0.06582674  0.12169604  0.07528304  0.07648554
  -0.30907994]
 [ 0.00205106  0.31606881  0.6262239   0.21417556 -0.14302547  0.15447466
   0.14917061  0.14121803 -0.17026002 -0.04962237  0.30769445 -0.49869142
  -0.02712539]
 [ 0.23932041 -0.0105905   0.61208035 -0.06085941  0.06610294 -0.10082451
   0.28696914 -0.09168285  0.42797018 -0.05574287 -0.20044931  0.47931378
   0.05279942]
 [-0.14199204  0.299634    0.13075693  0.35179658  0.72704851  0.03814394
  -0.3228833  -0.05677422 -0.15636143  0.06222011 -0.27140257  0.07128891
   0.06787022]
 [-0.39466085  0.06503951  0.14617896 -0.19806835 -0.14931841 -0.0841223
   0.02792498  0.46390791 -0.40593409 -0.30388245 -0.28603452  0.30434119
  -0.32013135]
 [-0.42293

In [11]:
print ("Eigenvalues", eigenvalues)

Eigenvalues [4.73243698 2.51108093 1.45424187 0.92416587 0.85804868 0.64528221
 0.55414147 0.10396199 0.35046627 0.16972374 0.29051203 0.22706428
 0.25232001]


It has been tested that eigenvectors have the same unit length for the axes therfore we can proceed with our assumption. The eigenvalues are now used to calculate the explained variance which tells us which principal component contributes how much to the variablility of the distribution. 

In [12]:
for i in sorted(eigenvalues):
    explained_variance = [(i / sum(eigenvalues))*100]
    print(explained_variance)

[0.7952148898994491]
[1.2982325756042155]
[1.7368356899899142]
[1.9300190939440784]
[2.2221534047897036]
[2.6807489483788647]
[4.238679322623324]
[4.935823319222569]
[6.563293679648605]
[7.069030182714031]
[11.123630536249992]
[19.207490257008928]
[36.198848099926316]


From the data given above we can see that first principal component explains 36.2% of the variability. The second data explains 19.2% of the variability. The third principal component explains 11.12% variability. Therefore we choose as many principal components so that the sum is more than 90% i.e more than 90% of the variablity in the dataset can be explained by the principal components.  

Adding up the 7 principal components we can say that 92% of the variability can be covered using 7 PC. 